### ISSUE TITLE
procstat metrics broken on FreeBSD since telegraf version 1.14.x and above

### ISSUE THREAD

**Issue by statseeker** [influxdata/telegraf#8062]
Hi Devs

The procstat metrics has been broken since 1.14.x and above on FreeBSD, it only provides the process name and not the metrics. Version 1.13.x and below works fine.

FreeBSD tested versions:

11.x

Uname:

FreeBSD 11.4-RELEASE-p2 FreeBSD 11.4-RELEASE-p2 #0: Thu Aug  6 03:03:06 UTC 2020     root@11_4-RELEASE:/usr/obj/usr/src/sys/GENERIC  amd64

Example from 1.14.x
./telegraf --version
Telegraf 1.14.0 (git: HEAD fefd7ff1)

command run:

./telegraf --config telegraf.conf --test

Basic config syntax:

[[inputs.procstat]]
  exe = "nim-snmptrapd"
[[inputs.procstat]]
  exe = "telegraf"
[[inputs.procstat]]
  exe = "nim-thresholds"
[[inputs.procstat]]
  exe = "rna-client-local"

Output from running a test:

> procstat_lookup,exe=nim-snmptrapd,host=qa-vm-back134.test.com,pid_finder=pgrep,result=success pid_count=1i,result_code=0i,running=0i 1598920068000000000
> procstat_lookup,exe=telegraf,host=qa-vm-back134.test.com,pid_finder=pgrep,result=success pid_count=0i,result_code=0i,running=0i 1598920068000000000
> procstat_lookup,exe=nim-thresholds,host=qa-vm-back134.test.com,pid_finder=pgrep,result=success pid_count=1i,result_code=0i,running=0i 1598920069000000000
> procstat_lookup,exe=rna-client-local,host=qa-vm-back134.test.com,pid_finder=pgrep,result=success pid_count=1i,result_code=0i,running=0i 1598920069000000000

Example from 1.13.4:
# ./telegraf --version
Telegraf 1.13.4 (git: HEAD ffabd6b5)

command run:

./telegraf --config telegraf.conf --test

Basic config syntax:

[[inputs.procstat]]
  exe = "nim-snmptrapd"
[[inputs.procstat]]
  exe = "telegraf"
[[inputs.procstat]]
  exe = "nim-thresholds"
[[inputs.procstat]]
  exe = "rna-client-local"

Output from running a test:

> processes,host=qa-vm-back134.test.com blocked=26i,idle=26i,running=2i,sleeping=44i,stopped=0i,total=99i,unknown=0i,wait=1i,zombies=0i 1598919972000000000
> procstat,exe=nim-snmptrapd,host=qa-vm-back134.test.com,process_name=nim-snmptrapd,user=testuser cpu_time_guest=0,cpu_time_guest_nice=0,cpu_time_idle=0,cpu_time_iowait=0,cpu_time_irq=0,cpu_time_nice=0,cpu_time_soft_irq=0,cpu_time_steal=0,cpu_time_system=81.648908,cpu_time_user=12.273174,cpu_usage=0.07376240212410023,memory_data=0i,memory_locked=0i,memory_rss=3469312i,memory_stack=0i,memory_swap=0i,memory_usage=0.027001256123185158,memory_vms=122904576i,num_threads=1i,pid=1882i,read_bytes=0i,read_count=1i,write_bytes=0i,write_count=10i 1598919972000000000
> procstat_lookup,exe=nim-snmptrapd,host=qa-vm-back134.test.com,pid_finder=pgrep,result=success pid_count=1i,result_code=0i,running=1i 1598919972000000000
> procstat_lookup,exe=telegraf,host=qa-vm-back134.test.com,pid_finder=pgrep,result=success pid_count=0i,result_code=0i,running=0i 1598919973000000000
> procstat,exe=nim-thresholds,host=qa-vm-back134.test.com,process_name=nim-thresholds,user=testuser cpu_time_guest=0,cpu_time_guest_nice=0,cpu_time_idle=0,cpu_time_iowait=0,cpu_time_irq=0,cpu_time_nice=0,cpu_time_soft_irq=0,cpu_time_steal=0,cpu_time_system=29.152165,cpu_time_user=7.178319,cpu_usage=0,memory_data=0i,memory_locked=0i,memory_rss=14168064i,memory_stack=0i,memory_swap=0i,memory_usage=0.11026841402053833,memory_vms=30121984i,num_threads=1i,pid=1931i,read_bytes=0i,read_count=4i,write_bytes=0i,write_count=50i 1598919973000000000
> procstat_lookup,exe=nim-thresholds,host=qa-vm-back134.test.com,pid_finder=pgrep,result=success pid_count=1i,result_code=0i,running=1i 1598919973000000000
> procstat,exe=rna-client-local,host=qa-vm-back134.test.com,process_name=rna-client-local,user=testuser cpu_time_guest=0,cpu_time_guest_nice=0,cpu_time_idle=0,cpu_time_iowait=0,cpu_time_irq=0,cpu_time_nice=0,cpu_time_soft_irq=0,cpu_time_steal=0,cpu_time_system=5.726136,cpu_time_user=1.543225,cpu_usage=0,memory_data=0i,memory_locked=0i,memory_rss=2891776i,memory_stack=0i,memory_swap=0i,memory_usage=0.022506359964609146,memory_vms=10141696i,num_threads=1i,pid=1697i,read_bytes=0i,read_count=0i,write_bytes=0i,write_count=14i 1598919974000000000

Thanks

**Comment by sjwang90** [comment]
Hi @statseeker, I realized no one has gotten to this. Are you still seeing these broken metrics?

**Comment by statseeker** [comment]
correct its still an issue

**Comment by ben-dl** [comment]
Still broken as of Telegraf 1.17.0.

**Comment by gaspar0069** [comment]
Using Telegraf 1.17.0 on FreeBSD 12.2, only had procstat_lookup_ metrics being populated.

Mounting the proc filesystem to /proc corrected the issue for me:

`mount -t procfs proc /proc`

Afterwards the procstat_ metrics appears to be being populated as expected.
(I'm just an end-user, so take this tidbit of info as a simple workaround or as a hint to figuring out the underlying cause for the change in behavior after 1.13.x on FreeBSD.)

**Comment by powersj** [comment]
Hi,

I've been trying to track down what changed, but v1.13 and v1.14 releases are very old and I really do not see any telegraf changes that affected procstat between those release branches.

> Afterwards the procstat_ metrics appears to be being populated as expected.

We didn't seem to have a go.mod with changes tracked back at this time. I do wonder if a change to gopsutil, the library used to collect this information, forced the need to have /proc mounted.

I wanted to reproduce this, but I see the same thing across versions:

```
powersj@:~ $ ./telegraf-1.13.4/usr/bin/telegraf --config config.toml --debug --test
2023-02-06T21:33:12Z I! Starting Telegraf 1.13.4
2023-02-06T21:33:12Z D! [agent] Initializing plugins
> procstat_lookup,exe=sshd,pid_finder=pgrep,result=success pid_count=0i,result_code=0i,running=0i 1675719193000000000

powersj@:~ $ ./telegraf-1.14.0/usr/bin/telegraf --config config.toml --debug --test
2023-02-06T21:33:04Z I! Starting Telegraf 1.14.0
2023-02-06T21:33:04Z D! [agent] Initializing plugins
> procstat_lookup,exe=sshd,pid_finder=pgrep,result=success pid_count=0i,result_code=0i,running=0i 1675719185000000000

powersj@:~ $ ./telegraf-1.25.1/usr/bin/telegraf --config config.toml --debug --test
2023-02-06T21:33:18Z I! Starting Telegraf 1.25.1
2023-02-06T21:33:18Z I! Available plugins: 212 inputs, 9 aggregators, 26 processors, 21 parsers, 57 outputs, 1 secret-stores
2023-02-06T21:33:18Z I! Loaded inputs: procstat
2023-02-06T21:33:18Z I! Loaded aggregators: 
2023-02-06T21:33:18Z I! Loaded processors: 
2023-02-06T21:33:18Z I! Loaded secretstores: 
2023-02-06T21:33:18Z W! Outputs are not used in testing mode!
2023-02-06T21:33:18Z I! Tags enabled: host=
2023-02-06T21:33:18Z D! [agent] Initializing plugins
2023-02-06T21:33:18Z D! [agent] Starting service inputs
2023-02-06T21:33:18Z D! [agent] Stopping service inputs
2023-02-06T21:33:18Z D! [agent] Input channel closed
2023-02-06T21:33:18Z D! [agent] Stopped Successfully
> procstat_lookup,exe=sshd,pid_finder=pgrep,result=success pid_count=0i,result_code=0i,running=0i 1675719199000000000
```

Am I missing something in order to reproduce this further?

**Comment by gaspar0069** [comment]
I'm currently on FreeBSD 13.1 and telegraf 1.24.1 from FreeBSD ports -- everything runs correctly without /proc mounted:

telegraf 1.24.1, proc is not mounted, both procstat and procstat_lookup are populated as normal:
```
$:/tmp# cat /tmp/telegraf.test
[[inputs.procstat]]
exe = "bacula-dir"

$:/tmp# telegraf --config telegraf.test --test
2023-02-06T23:55:33Z I! Starting Telegraf unknown
2023-02-06T23:55:33Z I! Available plugins: 208 inputs, 9 aggregators, 26 processors, 20 parsers, 57 outputs
2023-02-06T23:55:33Z I! Loaded inputs: procstat
2023-02-06T23:55:33Z I! Loaded aggregators: 
2023-02-06T23:55:33Z I! Loaded processors: 
2023-02-06T23:55:33Z W! Outputs are not used in testing mode!
2023-02-06T23:55:33Z I! Tags enabled: host=monolith
> procstat,exe=bacula-dir,host=monolith,process_name=bacula-dir,user=bacula cpu_time_guest=0,cpu_time_guest_nice=0,cpu_time_idle=0,cpu_time_iowait=0,cpu_time_irq=0,cpu_time_nice=0,cpu_time_soft_irq=0,cpu_time_steal=0,cpu_time_system=6.698472,cpu_time_user=18.639482,cpu_usage=0,created_at=1666756956920000000i,memory_data=0i,memory_locked=0i,memory_rss=12177408i,memory_stack=0i,memory_swap=0i,memory_usage=0.01776733063161373,memory_vms=39112704i,num_threads=3i,pid=21977i,ppid=1i,read_bytes=0i,read_count=115i,write_bytes=0i,write_count=3137i 1675727734000000000
> procstat_lookup,exe=bacula-dir,host=monolith,pid_finder=pgrep,result=success pid_count=1i,result_code=0i,running=1i 1675727734000000000
```


Now I went back to an old copy of telegraf 1.17, but still able to run on my current FreeBSD 13.1 and can reproduce the error when /proc is not mounted:
```
$:/usr/local/.zfs/snapshot/20210425-12.1p3-preupgrade/bin# ./telegraf version
Telegraf 1.17.0
$:/usr/local/.zfs/snapshot/20210425-12.1p3-preupgrade/bin# ./telegraf --config /tmp/telegraf.test --test
2023-02-07T00:01:02Z I! Starting Telegraf 1.17.0
> procstat_lookup,exe=bacula-dir,host=monolith,pid_finder=pgrep,result=success pid_count=1i,result_code=0i,running=0i 1675728063000000000
```
procstat_lookup finds the instance (pid_count=1) but procstat is not returned.

```
$:/usr/local/.zfs/snapshot/20210425-12.1p3-preupgrade/bin# mount -t procfs proc /proc
$:/usr/local/.zfs/snapshot/20210425-12.1p3-preupgrade/bin# mount | grep proc
procfs on /proc (procfs, local)
$:/usr/local/.zfs/snapshot/20210425-12.1p3-preupgrade/bin# ./telegraf --config /tmp/telegraf.test --test
2023-02-07T00:01:39Z I! Starting Telegraf 1.17.0
> procstat,exe=bacula-dir,host=monolith,process_name=bacula-dir,user=bacula cpu_time_guest=0,cpu_time_guest_nice=0,cpu_time_idle=0,cpu_time_iowait=0,cpu_time_irq=0,cpu_time_nice=0,cpu_time_soft_irq=0,cpu_time_steal=0,cpu_time_system=6.698472,cpu_time_user=18.640221,cpu_usage=0,memory_data=0i,memory_locked=0i,memory_rss=12177408i,memory_stack=0i,memory_swap=0i,memory_usage=0.01776733063161373,memory_vms=39112704i,num_threads=3i,pid=21977i,read_bytes=0i,read_count=115i,write_bytes=0i,write_count=3137i 1675728100000000000
> procstat_lookup,exe=bacula-dir,host=monolith,pid_finder=pgrep,result=success pid_count=1i,result_code=0i,running=1i 1675728100000000000
```
After mounting /proc both procstat_lookup and procstat return data.
```
$:/usr/local/.zfs/snapshot/20210425-12.1p3-preupgrade/bin# umount /proc
$:/usr/local/.zfs/snapshot/20210425-12.1p3-preupgrade/bin# ./telegraf --config /tmp/telegraf.test --test
2023-02-07T00:01:50Z I! Starting Telegraf 1.17.0
> procstat_lookup,exe=bacula-dir,host=monolith,pid_finder=pgrep,result=success pid_count=1i,result_code=0i,running=0i 1675728111000000000
```
And unmounting /proc again reproduces the lack of procstat data.

So it does seem like the issue has resolved somewhere between 1.17 and 1.24.1.

@powersj, looking at your example runs, it looks like your telegraf is not finding a sshd running, so procstat is not returned anyways (procstat_lookup ran "success"fully, but didn't find any exe=sshd, and returns a pid_count=0).

**Comment by powersj** [comment]
> looking at your example runs, it looks like your telegraf is not finding a sshd running

woops, good point. Switching to look for `sendmail` or `dhclient` worked out better to reproduce the issue on v1.14.0

> So it does seem like the issue has resolved somewhere between 1.17 and 1.24.1.

Thank you for confirming this and I note 1.25.1 works locally.

I am going to go ahead and close this with the note, that it appears later versions are working as intended.

Thanks again for the quick confirmation!

**Comment by linearray** [comment]
With telegraf-1.25.0_1 on FreeBSD 13.2 this is broken again, or it was a different issue in the first place.

`/proc` mounted or not, `procstat` will not do anything, despite `procstat_lookup` finding matching processes.

```
> telegraf --config conf.d/telegraf.conf --test | grep proc
2023-06-02T23:29:49Z I! Starting Telegraf unknown
2023-06-02T23:29:49Z I! Available plugins: 213 inputs, 9 aggregators, 26 processors, 21 parsers, 57 outputs, 1 secret-stores
2023-06-02T23:29:49Z I! Loaded inputs: cpu disk diskio mem net netstat processes procstat system
2023-06-02T23:29:49Z I! Loaded aggregators: 
2023-06-02T23:29:49Z I! Loaded processors: 
2023-06-02T23:29:49Z I! Loaded secretstores: 
2023-06-02T23:29:49Z W! Outputs are not used in testing mode!
2023-06-02T23:29:49Z I! Tags enabled: host=influxdb-live-0
> disk,device=procfs,fstype=procfs,host=influxdb-live-0,mode=rw,path=/proc free=0i,inodes_free=0i,inodes_total=1i,inodes_used=1i,total=4096i,used=4096i,used_percent=100 1685748590000000000
> disk,device=procfs,fstype=procfs,host=influxdb-live-0,mode=rw,path=/proc free=0i,inodes_free=0i,inodes_total=1i,inodes_used=1i,total=4096i,used=4096i,used_percent=100 1685748590000000000
> processes,host=influxdb-live-0 blocked=13i,idle=25i,running=5i,sleeping=15i,stopped=0i,total=59i,unknown=0i,wait=1i,zombies=0i 1685748590000000000
> procstat_lookup,host=influxdb-live-0,pattern=java,pid_finder=pgrep,result=success pid_count=1i,result_code=0i,running=0i 1685748590000000000
```

**Comment by gaspar0069** [comment]
I'm currently using telegraf 1.25.0_1, but am still on 13.1, just as a data point that that combo still works (without proc mounted).

One commonality I noticed between @linearray 's failure test case and my own on 1.17.0 was that although
`procstat_lookup` had `pid_count=1i`, running was `running=0i`, whereas in the successful runs that populated `procstat` (1.17.0 w/ proc mounted or 1.25.0_1 on FreeBSD 13.1), running was `running=1i`.

I'm not familiar with telegraf code at all, I'm just an end-user, but perhaps this suggests a failure somewhere between finding a pid and determining that the proper process is running so that it can grab data from it.

@linearray : Just to rule out any strange setting that might affect results, could you post the relevant `[[inputs.procstat]]` of your telegraf.conf?  Can you confirm that procstat fails on any other running process you try to look up, not just the "java" one?

**Comment by linearray** [comment]
Good catch with the `running=0i`. I will see what I can dig up.

`telegraf.conf`:
```
[[inputs.procstat]]
  pattern = "mg"
```

Result is the same regardless of pattern.

**Comment by linearray** [comment]
By chance I noticed `procstat` metrics showing up in influxdb. I investigated further and noticed that all the reporting nodes are amd64, while all the nodes where `procstat` fails are arm64.

Indeed the outputs I posted above are from an arm64 machine, while https://github.com/influxdata/telegraf/issues/8062#issuecomment-1581292614 is from an amd64 machine, so architecture seems to play a role in this.
